atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/LockWatchValueCacheImpl.java

```diff
@@ -0,0 +1,168 @@
+    //  issue. Chances are that this may need to be re-jigged to take a batch, and be connected to the batched commit
+    //  timestamp call.
+    @Override
+    public synchronized void updateCache(TransactionDigest digest, long startTs) {
+        try {
+            CommitUpdate commitUpdate = eventCache.getCommitUpdate(startTs);
+            commitUpdate.accept(new Visitor<Void>() {
+                @Override
+                public Void invalidateAll() {
+                    // If this is an election, we should throw. If not and we are not a serialisable transaction, we are
+                    // *technically* ok to go on here. However, we cannot determine which of the two we are in, so we
+                    // throw anyway.
+                    throw new TransactionLockWatchFailedException("A Timelock leader election has occurred between the"
+                        + " start and commit time of this transaction, and thus all events have been invalidated. This"
+                        + " transaction will be retried");
+                }
+
+                @Override
+                public Void invalidateSome(Set<LockDescriptor> invalidatedLocks) {
+                    Set<CellReference> invalidatedCells = invalidatedLocks.stream()
+                            .flatMap(LockWatchValueCacheImpl::extractTableAndCell)
+                            .collect(Collectors.toSet());
+                    KeyedStream.stream(digest.loadedValues())
+                            .filterKeys(cellReference -> !invalidatedCells.contains(cellReference))
+                            .forEach(valueStore::putValue);
+                    return null;
+                }
+            });
+        } finally {
+            snapshotStore.removeTimestamp(StartTimestamp.of(startTs));
+        }
+    }
+
+    @Override
+    public TransactionScopedCache createTransactionScopedCache(long startTs) {
+        // todo(jshah): implement
+        throw new UnsupportedOperationException("Not yet implemented");
+    }
+
+    /**
+     * In order to maintain the necessary invariants, we need to do the following:
+     *
+     *  1. For each new event, we apply it to the cache. The effects of this application is described in
+     *     {@link LockWatchValueCache}.
+     *  2. For each transaction, we must ensure that we store a snapshot of the cache at the sequence corresponding
+     *     to the transaction's start timestamp. Note that not every sequence will have a corresponding timestamp, so we
+     *     don't bother storing a snapshot for those sequences. Also note that we know that each call here will only
+     *     ever have new events, and that consecutive calls to this method will *always* have increasing sequences
+     *     (without this last guarantee, we'd need to store snapshots for all sequences).
+     */
+    private void updateStores(TransactionsLockWatchUpdate updateForTransactions) {
```

Comment:
I will be extra defensive against this, but it turns out that this is already defended against in LWEC - if you try and get updates for transactions that have a latest version V, and you are on C > V, it will throw. I'm also pretty certain that it only returns updates between C+1 and V inclusive when C < V, so I think we're actually safe here because of the defensiveness of LWEC.